examples/hamiltonian_representation.py

```diff
@@ -0,0 +1,235 @@
+        return hamiltonian
+
+    raise ValueError(f'Unsupported type: {type(boolean_expr)}')
+
+
+def get_name_to_id(boolean_exprs: Sequence[Expr]) -> Dict[str, int]:
+    """Maps the variables to a unique integer.
+
+    Args:
+        boolean_expr: A Sympy expression containing symbols and Boolean operations
+
+    Return:
+        A dictionary of string (the variable name) to a unique integer.
+    """
+
+    # For run-to-run identicalness, we sort the symbol name lexicographically.
+    symbol_names = sorted(
+        {symbol.name for boolean_expr in boolean_exprs for symbol in boolean_expr.free_symbols}
+    )
+    return {symbol_name: i for i, symbol_name in enumerate(symbol_names)}
+
+
+def _gray_code_comparator(k1, k2, flip=False):
+    max_1 = k1[-1] if k1 else -1
+    max_2 = k2[-1] if k2 else -1
+    if max_1 != max_2:
+        return -1 if (max_1 < max_2) ^ flip else 1
+    if max_1 == -1:
+        return 0
+    return _gray_code_comparator(k1[0:-1], k2[0:-1], not flip)
+
+
+def build_circuit_from_hamiltonians(
+    hamiltonian_lists: List[HamiltonianList], qubits: List[cirq.NamedQubit], theta: float
+) -> cirq.Circuit:
+    """Builds a circuit according to [1].
+
+    Args:
+        hamiltonian_lists: the list of Hamiltonians, typically built by calling
+            build_hamiltonian_from_boolean().
+        qubits: The list of qubits corresponding to the variables.
+        theta: A single float scaling the rotations.
+
+    Return:
+        A dictionary of string (the variable name) to a unique integer.
+    """
+    combined = HamiltonianList.O()
+    for hamiltonian_list in hamiltonian_lists:
+        combined += hamiltonian_list
+
+    circuit = cirq.Circuit()
+
+    # Here we follow improvements of [4] cancelling out the CNOTs. The first step is to order by
+    # Gray code so that as few as possible gates are changed.
+    sorted_hs = sorted(
+        list(combined.hamiltonians.keys()), key=functools.cmp_to_key(_gray_code_comparator)
+    )
+
+    # Applies the CNOTs
+    def _apply_cnots(h):
+        circuit.append([cirq.CNOT(qubits[c], qubits[h[-1]]) for c in h[0:-1]])
+
+    previous_h: Tuple[int, ...] = ()
+    for h in sorted_hs:
+        w = combined.hamiltonians[h]
+
+        # We first test whether the rotation is applied on the same qubit.
+        last_qubit_is_same = previous_h and h and previous_h[-1] == h[-1]
+        if last_qubit_is_same:
+            # Instead of applying previous_h and then h, we just apply the symmetric difference of
+            # the two CNOTs.
+            h_diff = tuple(sorted(set(previous_h).symmetric_difference(h)))
+            _apply_cnots(h_diff + (h[-1],))
+            # This would be equivalent to the lines below, but it should use fewer gates.
+        else:
+            _apply_cnots(previous_h)
+            _apply_cnots(h)
+
+        if len(h) >= 1:
+            circuit.append(cirq.Rz(rads=(theta * w)).on(qubits[h[-1]]))
+
+        previous_h = h
+
```

Comment:
I really like this change combining two ideas. But compared with the previous code, this version becomes less readable. (Mainly because here it is always from [0 to 2^n-1], the structure of gray code is not traversed.) So I suggest keep the older structure:
```
for h in sorted_hs:
      _apply_cnots(previous_h, h, forward=True)
      if len(h) >= 1:
            circuit.append(cirq.Rz(rads=(theta * w)).on(qubits[h[-1]]))
      previous_h = h
```
move the detailed cnot canceling into sub-functions

Reply:
Done. I wasn't sure what the `forward` was for though, but otherwise I did what you suggested. It's much easier to read indeed.